Always add decimal when converting float to string

KoBeWi opened the pull request:
So that
```
var a: float = 1
print(a)
```
prints
```
1.0
```
Makes working with numbers less confusing. I remember someone mentioned it around https://github.com/godotengine/godot-proposals/issues/1866

Closes https://github.com/godotengine/godot-proposals/issues/7894

Calinou commented:
Does this change apply automatically to floats within Vectors/Colors as well?

KoBeWi commented:
Hmm, it doesn't. Changing it might be not as easy :I

KoBeWi commented:
Ok, so I extended this to these types (in addition to floats): Vector2, Rect2, Vector3.
The original aim was to differentiate integer types from floats. As we don't have AABBi or Basisi I left these types untouched.

dalexeev commented:
Also need to solve this problem:

```
print(5.0)                    # 5
print(5.0000001234)           # 5 <--
print("%.16f" % 5.0)          # 5.0000000000000000
print("%.16f" % 5.0000001234) # 5.0000001234000004
```

```
$ python3 -q
>>> print(5.0000001234)
5.0000001234
```

KoBeWi commented:
The precision problem doesn't look related.

dalexeev commented:
> The precision problem doesn't look related.

The problem is not with the precision of the numbers themselves, but that `print` doesn't display the number as precisely as it could. This can be seen by the fact that everything works with the formatting operator.

KoBeWi commented:
Ah, right. I don't know how to fix this tbh. I can't use `is_equal_approx()` in FLOAT_STRING, because it produces stuff like `1.00001.0`.

KoBeWi commented:
Ok changed it. Here's the FLOAT_STRING as it was originally:
```
#define FLOAT_STRING(f) (rtos(f) + (Math::floor(f) == f ? ".0" : ""))
```
The new one is slightly less efficient, but maybe it doesn't matter that much.

akien-mga commented:
We approved the change in a PR review meeting today, though we were concerned about the performance cost of the `FLOAT_STRING` define (doing a costly calculation twice + a `find()`).

A better option might be to use `String::num_real()` instead of `String::num()`, which already takes care of adding a `.0` where relevant. I'm not sure how `num()` and `num_real()` differ aside from that and whether there would be some stuff handled by the one and not the other.

KoBeWi commented:
Rebased and reworked. I removed the macro and just changed usages of `String::num_real()` to use `true` as second argument (which makes the `.0` always appear). I did it in the operators instead of `stringify()` method to avoid duplicating formats. As a result, some more classes are affected, e.g. Basis, which uses vectors.

```
print(1.0)
print(Vector2.ONE)
print(Rect2(Vector2.DOWN, Vector2.ONE))
print(Basis())
```
->
```
1.0
(1.0, 1.0)
[P: (0.0, 1.0), S: (1.0, 1.0)]
[X: (1.0, 0.0, 0.0), Y: (0.0, 1.0, 0.0), Z: (0.0, 0.0, 1.0)]
```

KoBeWi commented:
So I just noticed that changing `Variant::stringify()` will affect generated docs (as seen in failed CI, PI and TAU lose their precision). Not sure if that's ok 🤔

akien-mga commented on modules/gdscript/tests/scripts/parser/features/nested_arithmetic.out:
```diff
@@ -76,7 +76,7 @@ GDTEST_OK
 -2.718
 
-36900.9009009009
+36900.9
```
Seems like `num_real` hardcodes 6 "decimals" which in practice is 6 significant digits, so 36900.9.

`num` by default prints all the significant digits it can. IMO the `num_real` logic is needlessly restrictive.

akien-mga replied:
This shouldn't be blocking for this PR though but we need to look into it further, I think the hardcoded 6 doesn't make much sense especially since now `float` in GDScript in 64-bit.

akien-mga commented:
This seems to break the generated Mono glue:
```
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/VisualScript.cs(125,84): error CS1503: Argument 1: cannot convert from 'double' to 'float' [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/VisualScript.cs(125,89): error CS1503: Argument 2: cannot convert from 'double' to 'float' [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/TextParagraph.cs(351,110): error CS1503: Argument 1: cannot convert from 'double' to 'float' [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/TextParagraph.cs(351,115): error CS1503: Argument 2: cannot convert from 'double' to 'float' [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
```
CC @godotengine/mono

aaronfranke commented on core/variant/variant.cpp:
```diff
@@ -1664,7 +1664,7 @@ String Variant::stringify(int recursion_count) const {
 			return itos(_data._int);
 		case FLOAT:
-			return rtos(_data._float);
+			return String::num_real(_data._float, true);
```
`rtos` is designed to print out doubles (14 digits), while `String::num_real` is designed to print out `real_t`s (either 6 or 14 digits) for the types like Vector3. As-is this change isn't good for Variant because it uses double.

KoBeWi replied:
I changed how this method is implemented. You need to pass actual `real_t` to get the old behavior.

aaronfranke reviewed (commented):
For the Mono bindings generator, we'll probably need to manually generate each piece of the constructor.

neikeq commented:
The C# bindings generator can be fixed by appending the `f` suffix.

KoBeWi commented:
If this implementation is causing problems, I could go back to what I did before, i.e. change formatting for each type individually; it would be less intrusive. Should I change it?

akien-mga commented:
> If this implementation is causing problems, I could go back to what I did before, i.e. change formatting for each type individually; it would be less intrusive. Should I change it?

I think this is still the right approach, we just need to fix:

- Mono bindings to not rely on `operator String()` but another function / have their own conversion method to ensure that floats get the `f` suffix. That being said I'm puzzled that it works currently, as we do have some default values such as `gravity` with `Vector3(0, -9.8, 0)`. Was it just by luck that currently no default value with non-whole number is used in Mono bindings? So this sounds to me like a pre-existing bug.
- Change the behavior of `num` to do something similar to `num_real`, or add a dedicated method, so that explicit `double`s can still be printed properly.

neikeq commented:
I can have a look at this now, or we can push the fix after this is merged if you prefer that.

> Was it just by luck that currently no default value with non-whole number is used in Mono bindings? So this sounds to me like a pre-existing bug.

Indeed. To be fair, it was a quite safe assumption to make 😅

akien-mga commented:
> I can have a look at this now, or we can push the fix after this is merged if you prefer that.

Feel free to work on it now, this PR still needs some work to figure out the other problem with float/double pretty printing.

neikeq commented:
I pushed the fix :+1:

KoBeWi commented:
Something broke in the bindings after rebase.

raulsntos commented:
There was a mix-up in the NodePath ctor, I also took the opportunity to add the fix for the new Vector4, Vector4i and Projection types that were added after **neikeq**'s push.

```diff
diff --git a/modules/mono/editor/bindings_generator.cpp b/modules/mono/editor/bindings_generator.cpp
index 9855dc82c6..7c3fa28ef9 100644
--- a/modules/mono/editor/bindings_generator.cpp
+++ b/modules/mono/editor/bindings_generator.cpp
@@ -3272,6 +3272,13 @@ static String _get_vector3_cs_ctor_args(const Vector3 &p_vec3) {
 			String::num_real(p_vec3.z, true) + "f";
 }
 
+static String _get_vector4_cs_ctor_args(const Vector4 &p_vec4) {
+	return String::num_real(p_vec4.x, true) + "f, " +
+			String::num_real(p_vec4.y, true) + "f, " +
+			String::num_real(p_vec4.z, true) + "f, " +
+			String::num_real(p_vec4.w, true) + "f";
+}
+
 static String _get_vector2i_cs_ctor_args(const Vector2i &p_vec2i) {
 	return itos(p_vec2i.x) + ", " + itos(p_vec2i.y);
 }
@@ -3280,6 +3287,10 @@ static String _get_vector3i_cs_ctor_args(const Vector3i &p_vec3i) {
 	return itos(p_vec3i.x) + ", " + itos(p_vec3i.y) + ", " + itos(p_vec3i.z);
 }
 
+static String _get_vector4i_cs_ctor_args(const Vector4i &p_vec4i) {
+	return itos(p_vec4i.x) + ", " + itos(p_vec4i.y) + ", " + itos(p_vec4i.z) + ", " + itos(p_vec4i.w);
+}
+
 static String _get_color_cs_ctor_args(const Color &p_color) {
 	return String::num(p_color.r, 4) + ", " + String::num(p_color.g, 4) + ", " +
 			String::num(p_color.b, 4) + ", " + String::num(p_color.a, 4);
@@ -3316,7 +3327,7 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
 		case Variant::NODE_PATH:
 			if (r_iarg.type.cname == name_cache.type_StringName || r_iarg.type.cname == name_cache.type_NodePath) {
 				if (r_iarg.default_argument.length() > 0) {
-					r_iarg.default_argument = "(%s)\"" + p_val.default_argument + "\"";
+					r_iarg.default_argument = "(%s)\"" + p_val.operator String() + "\"";
 					r_iarg.def_param_mode = ArgumentInterface::NULLABLE_REF;
 				} else {
 					// No need for a special `in` statement to change `null` to `""`. Marshaling takes care of this already.
@@ -3375,8 +3386,11 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
 			r_iarg.def_param_mode = ArgumentInterface::NULLABLE_VAL;
 			break;
 		case Variant::VECTOR4:
+			r_iarg.default_argument = "new Vector4(" + _get_vector4_cs_ctor_args(p_val.operator Vector4()) + ")";
+			r_iarg.def_param_mode = ArgumentInterface::NULLABLE_VAL;
+			break;
 		case Variant::VECTOR4I:
-			r_iarg.default_argument = "new %s" + r_iarg.default_argument;
+			r_iarg.default_argument = "new Vector4i(" + _get_vector4i_cs_ctor_args(p_val.operator Vector4i()) + ")";
 			r_iarg.def_param_mode = ArgumentInterface::NULLABLE_VAL;
 			break;
 		case Variant::OBJECT:
@@ -3451,7 +3465,11 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
 			if (projection == Projection()) {
 				r_iarg.default_argument = "Projection.Identity";
 			} else {
-				r_iarg.default_argument = "new Projection(new Vector4" + projection.columns[0].operator String() + ", new Vector4" + projection.columns[1].operator String() + ", new Vector4" + projection.columns[2].operator String() + ", new Vector4" + projection.columns[3].operator String() + ")";
+				r_iarg.default_argument = "new Projection(new Vector4(" +
+						_get_vector4_cs_ctor_args(projection.columns[0]) + "), new Vector4(" +
+						_get_vector4_cs_ctor_args(projection.columns[1]) + "), new Vector4(" +
+						_get_vector4_cs_ctor_args(projection.columns[2]) + "), new Vector4(" +
+						_get_vector4_cs_ctor_args(projection.columns[3]) + "))";
 			}
 			r_iarg.def_param_mode = ArgumentInterface::NULLABLE_VAL;
 		} break;

```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3445,48 +3484,70 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
 		case Variant::RECT2I: {
 			Rect2i rect = p_val.operator Rect2i();
-			r_iarg.default_argument = "new Rect2I(new Vector2I" + rect.position.operator String() + ", new Vector2I" + rect.size.operator String() + ")";
+			r_iarg.default_argument = "new Rect2i(new Vector2i" +
+					_get_vector2i_cs_ctor_args(rect.position) + ", new Vector2i" +
+					_get_vector2i_cs_ctor_args(rect.size) + ")";
```
```suggestion
			r_iarg.default_argument = "new Rect2I(new Vector2I(" +
					_get_vector2i_cs_ctor_args(rect.position) + "), new Vector2I(" +
					_get_vector2i_cs_ctor_args(rect.size) + "))";
```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3445,48 +3484,70 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
+			r_iarg.def_param_mode = ArgumentInterface::NULLABLE_VAL;
+			break;
 		case Variant::VECTOR2I:
+			r_iarg.default_argument = "new Vector2i(" + _get_vector2i_cs_ctor_args(p_val.operator Vector2i()) + ")";
```
```suggestion
			r_iarg.default_argument = "new Vector2I(" + _get_vector2i_cs_ctor_args(p_val.operator Vector2i()) + ")";
```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3445,48 +3484,70 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
+			break;
 		case Variant::VECTOR3I:
-			r_iarg.default_argument = "new %s" + r_iarg.default_argument;
+			r_iarg.default_argument = "new Vector3i(" + _get_vector3i_cs_ctor_args(p_val.operator Vector3i()) + ")";
```
```suggestion
			r_iarg.default_argument = "new Vector3I(" + _get_vector3i_cs_ctor_args(p_val.operator Vector3i()) + ")";
```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3445,48 +3484,70 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
+			break;
 		case Variant::VECTOR4I:
-			r_iarg.default_argument = "new %s" + r_iarg.default_argument;
+			r_iarg.default_argument = "new Vector4i(" + _get_vector4i_cs_ctor_args(p_val.operator Vector4i()) + ")";
```
```suggestion
			r_iarg.default_argument = "new Vector4I(" + _get_vector4i_cs_ctor_args(p_val.operator Vector4i()) + ")";
```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3445,48 +3484,70 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
 		case Variant::AABB: {
 			AABB aabb = p_val.operator ::AABB();
-			r_iarg.default_argument = "new Aabb(new Vector3" + aabb.position.operator String() + ", new Vector3" + aabb.size.operator String() + ")";
+			r_iarg.default_argument = "new AABB(new Vector3" +
+					_get_vector3_cs_ctor_args(aabb.position) + ", new Vector3" +
+					_get_vector3_cs_ctor_args(aabb.size) + ")";
```
```suggestion
			r_iarg.default_argument = "new Aabb(new Vector3(" +
					_get_vector3_cs_ctor_args(aabb.position) + "), new Vector3(" +
					_get_vector3_cs_ctor_args(aabb.size) + "))";
```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3417,9 +3417,43 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
+}
+
+static String _get_color_cs_ctor_args(const Color &p_color) {
+	return String::num(p_color.r, 4) + ", " + String::num(p_color.g, 4) + ", " +
+			String::num(p_color.b, 4) + ", " + String::num(p_color.a, 4);
```
Why are the vector arguments using `String::num_real` but Color uses `String::num`?

Also, we currently don't have any default Color arguments with decimals, but to future proof it we should probably add the `f` suffix here.

```suggestion
	return String::num(p_color.r, 4) + "f, " +
			String::num(p_color.g, 4) + "f, " +
			String::num(p_color.b, 4) + "f, " +
			String::num(p_color.a, 4) + "f";
```

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -3445,48 +3484,70 @@ bool BindingsGenerator::_arg_default_value_from_variant(const Variant &p_val, Ar
 		case Variant::PLANE: {
 			Plane plane = p_val.operator Plane();
-			r_iarg.default_argument = "new Plane(new Vector3" + plane.normal.operator String() + ", " + rtos(plane.d) + ")";
+			r_iarg.default_argument = "new Plane(new Vector3(" +
+					_get_vector3_cs_ctor_args(plane.normal) + "), " + rtos(plane.d) + ")";
```
We currently don't seem to have any default Plane arguments, but to future proof it, we probably should add the `f` suffix to the point argument.

```suggestion
			r_iarg.default_argument = "new Plane(new Vector3(" +
					_get_vector3_cs_ctor_args(plane.normal) + "), " + rtos(plane.d) + "f)";
```

raulsntos reviewed (commented):
Suggestions to fix the .NET bindings generator.

KoBeWi commented on modules/gdscript/tests/scripts/parser/features/number_separators.out:
```diff
@@ -11,14 +11,14 @@ GDTEST_OK
--1234.4567
--1234
--1234
+-1234.457
+-1234.457
```
Something wrong with precision here 🤔

KoBeWi committed:
Always add decimal when printing float

KoBeWi commented on core/extension/extension_api_dump.cpp:
```diff
@@ -1360,6 +1360,9 @@ static bool compare_dict_array(const Dictionary &p_old_api, const Dictionary &p_
 		ERR_FAIL_COND_V_MSG(!elem.has(p_name_field), false, vformat("Validate extension JSON: Element of base_array '%s' is missing field '%s'. This is a bug.", base_array, p_name_field));
 		String name = elem[p_name_field];
+		if (name.is_valid_float()) {
+			name = name.trim_suffix(".0"); // Make "integers" stringified as integers.
```
Thanks JSON 😬

Calinou reviewed (commented):
Tested locally, it works as expected with `float` but not `Vector2` (tested on single-precision build).

```gdscript
func _ready() -> void:
	print(-0.0 * Vector2.ONE)
	print(-0.0000000001 * Vector2.ONE)
	print(-1.0 * Vector2.ONE)
	print(-1.0000000001 * Vector2.ONE)
	print(0.0 * Vector2.ONE)
	print(0.0000000001 * Vector2.ONE)
	print(1.0 * Vector2.ONE)
	print(1.0000000001 * Vector2.ONE)
```

Results in:

```
(0.0, 0.0)
(-0, -0)
(-1.0, -1.0)
(-1.0, -1.0)
(0.0, 0.0)
(0, 0)
(1.0, 1.0)
(1.0, 1.0)
```

In `master`, this is the output:

```
(0, 0)
(-0, -0)
(-1, -1)
(-1, -1)
(0, 0)
(0, 0)
(1, 1)
(1, 1)
```

Notice the `0` and `-0` which shouldn't be here, as Vector2 is always floating-point.

The same occurs with `Vector3`.

While this issue is not introduced by this PR, I think people will have more confidence in the displayed type if we merge this PR. This means that we need to get the displayed type right, as showing a misleading type will be a lot more problematic once this is merged.

KoBeWi commented:
Fixed. `String::num()` was removing all trailing zeroes.

Calinou reviewed (approved):
Tested locally, it works as expected with vector types now:

```
(0.0, 0.0)
(-0.0, -0.0)
(-1.0, -1.0)
(-1.0, -1.0)
(0.0, 0.0)
(0.0, 0.0)
(1.0, 1.0)
(1.0, 1.0)
```

It still says `-0.0` on the second line, but this is technically more correct given the number is being rounded from a very low negative value. (This rounding should be addressed in a separate PR.)

The PR also affects the inspector correctly as well now:

![image](https://github.com/godotengine/godot/assets/180032/b1d0e61c-260a-4c3a-8566-18171d1c369d)

m4rr5 commented:
Since this PR ended up improving the output of String::num() I noticed that this same code also reports back if a float/double is "nan" or "inf". For "inf" it also checks, using signbit(), if it's "+inf" or "-inf". I was wondering if it makes sense to also do this for zero values and nan values. I don't think this is just a theoretical improvement either. Let me give an example I bumped into recently. When you calculate the "arc tangent" of a number, an often used way to do that is using some implementation (see for example https://en.cppreference.com/w/cpp/numeric/math/atan2) of atan2(y, x). There are edge cases where this function will return a different value based on if one of the inputs was +0 or -0, therefore I think it makes sense if printing such a value would show the sign bit correctly. I'm not saying this should be part of this PR since it addresses a different issue, it's just that me reading this triggered the thought.

aaronfranke commented:
This PR breaks line numbers in the script editor. Line numbers are integers, they should not have `.0` at the end.
<img width="129" alt="Screenshot 2024-04-19 at 2 49 47 PM" src="https://github.com/godotengine/godot/assets/1646875/544330eb-b9e9-4b60-9735-2095c4194610">

I tested this on [GodSVG](https://github.com/MewPurPur/GodSVG), a project that heavily relies on float stringification, and here are the results:

<img width="400" alt="Screenshot 2024-04-19 at 2 33 54 PM" src="https://github.com/godotengine/godot/assets/1646875/88e07bc7-8dd1-4a9d-87f3-538dfce9cf43"> <img width="400" alt="Screenshot 2024-04-19 at 2 27 40 PM" src="https://github.com/godotengine/godot/assets/1646875/fe8a39de-ab26-42f8-a8e6-4d3dc3b9f4cc">

The resulting SVG text looks quite different, but it is actually completely valid, just not the most minimal. This is a good sign that even in UI-heavy applications, this PR will not break things. GodSVG already has code to minify floats, which is why `0.0` becomes `.0`, because it expects any number with `.` is not a whole number. The float minifier will only require small tweaks to account for the trailing `.0`. But anyway, the project still works correctly exactly as-is.

KoBeWi commented:
> This PR breaks line numbers in the script editor. Line numbers are integers, they should not have .0 at the end.

I changed `String::num()` to add trailing zero, because it takes `double` as argument. Seems like CodeEdit uses this method.

aaronfranke reviewed (approved):
I tested this on The Mirror and the app still works fine, and in fact it highlighted a pre-existing bug which I've fixed here https://github.com/the-mirror-gdp/the-mirror/pull/115

I am clicking approve because I support this change, the code looks good from a quick review, and I have tested it on 2 large projects. However, I am approving it on the condition that this is merged immediately after the release of Godot 4.3, so that it can get lots of testing in master before being included in Godot 4.4, since this PR changes a fundamental part of the engine and is inherently risky it needs testing. I also removed the `cherrypick:3.x` label for the same reason.

Mickeon reviewed (approved):
Remember to make a big note on this on the next changelog.

Repiteo commented:
Thanks!

Lippanon commented:
Does this only affect `print` or does it also change the behaviour of `str` in @GlobalScope? I relied on it to neatly display values when there were no decimal cases. It behaved exactly how I wanted, so is there no option to choose the behaviour? 
What's the alternative now? Having to run `step_decimals` on every float value to determine wether to use "%.0f" or not?

Bromeon commented:
Was there some research that went into this decision? The initial post has lots of upvotes, but it's naturally biased because everyone happy with the existing behavior would not come here.

In my opinion, this is a significant breaking change and the downsides should be considered carefully. The picture example is quite compelling here, it makes things harder to read and edit in cases where integers are needed.

Furthermore, while Python does it, I'm not sure if that's so relevant? GDScript syntactically resembles it, but for Godot gamedev other languages such as C# and C++ (or even Rust) are more relevant, and they all format floats without `.0` by default.

Not saying there are no use cases for this, but besides "is it useful", we should probably consider practical implications, namely:
1. Should the default stay as-is and this new behavior be opt-in?
1. If not, is there an easy way to emulate the old behavior? What do we tell users whose code is now broken?
1. Independent of the above, can we let UI elements like matrices/vectors still use the old format for readability/editability purposes?

clayjohn commented:
CC @aaronfranke @Calinou @Mickeon

Mickeon commented:
> Was there some research that went into this decision? The initial post has lots of upvotes, but it's naturally biased because everyone happy with the existing behavior would not come here.

It was a relatively brief, final discussion about it, but the PR had a whooping 3 years of time to cook. The consensus is that this should be merged as **early** as possible in the development cycle so that feedback can be gathered and evaluated as soon as possible, instead of breaking everything late when no one has time to adapt.

I myself anticipated this would have unforeseen consequences like the above. This is exactly what we need, and I appreciate it. Now the question is how to go about it.

> Should the default stay as-is and this new behavior be opt-in?

Unsure of how. It only sounds feasible as a project setting, which sounds like a bad idea, as it would influence the result of all GDScript code.

> If not, is there an easy way to emulate the old behavior? What do we tell users whose code is now broken?

For most cases, I can imagine that casting the float to an integer before displaying it would fix this, if the value would be equivalent.

> Independent of the above, can we let UI elements like matrices/vectors still use the old format for readability/editability purposes?

Perhaps there could be toggle in SpinBox itself for this, although it sounds somewhat specific.

aaronfranke commented:
@Lippanon As I noted above, projects may need to change their code if they want to keep the same behavior, especially GodSVG. However, GodSVG keeps working, so amazingly, it didn't fully break compatibility, it just changed the visual appearance and resulted in sub-optimal SVG files, which will be easy for @MewPurPur to fix, and easy for other devs in their projects. The upside is huge: floats are stringified as float literals by default, making the type clear.

Lippanon commented:
You claim it's easy yet provided no clean alternative. For example an RPG character page or dynamic tooltip for a skill or unit in an RTS, one can't simply cast to int, and having to call a function that checks the decimal cases and attributes the correct string formatting in gdscript is, in my opinion, not an acceptable alternative for a behaviour that has been the same for many years (if not since Godot's inception) and changed due a proposal that had ~25 thumbs-up (as mentioned by Bromeon nobody who likes the current behaviour will go out of their way to look for something that changes to disagree with it, and most users don't have time to check every proposal or PR).
I'll happily support the inclusion of this feature/new behaviour, but I don't support forcing such a change onto users. The only issue I see linked was [this](https://github.com/godotengine/godot-proposals/issues/1693) and their main problem was not being able to have sliders with integers.
The inspector in the editor seems cluttered to me now, harder to read, besides the aforementioned problem.

aaronfranke commented:
We absolutely should provide a function for the old behavior. It would be a static method in String, `num_something`.

> changed due a proposal that had ~25 thumbs-up (as mentioned by Bromeon nobody who likes the current behaviour will go out of their way to look for something that changes to disagree with it, and most users don't have time to check every proposal or PR).

Users can retroactively leave thumbs-downs if they disagree with the change. While votes are not guaranteed to decide what we go with, it is still extremely useful to be able to measure user feedback in this way.

Calinou commented:
> If not, is there an easy way to emulate the old behavior? What do we tell users whose code is now broken?

`str(some_integer).trim_suffix(".0")` should do the trick.

fire commented:
I am trending towards using the old behaviour, adding a new method for the new behaviour, and making each change case by case to the latest behaviour to keep backwards compatibility.

KoBeWi commented:
> would it be possible to have a setting on Label/RichTextLabel that toggles this change on/off for floats set as text

No. You can't set float to a Label, only set the final String.

We can add a utility method that converts float to String and trims zero when necessary.

dmbro commented:
> No. You can't set float to a Label, only set the final String.

Oh yeah, duh haha.

> We can add a utility method that converts float to String and trims zero when necessary.

Would that look something like `float_to_str(float_value)` or `str(float_value).trim_zero`?

aaronfranke commented:
Regardless, we should provide methods for each style of behavior. Users shouldn't need to write `str(float_value).trim_zero()`.

Would anyone like to propose APIs for these cases?

* If this PR is kept, what should the non-`.0` function be called?
* If this PR is reverted, what should the `.0` function be called?

PraxisMapper commented:
I think this change causes my project that runs fine in 4.3 to throw errors in 4.4 dev4.

I have several data files generated by a separate C# app, where keys in a dictionary are ints that can be referenced by other data files. The trick is that C# exports JSON dictionaries with int keys as strings instead, and I have not found a way to disable/change that behavior, so when I go to do that lookup I end up with something like 
```DataDict[str(RefKey)]```
which works fine in 4.3. In 4.4 dev4, these need to be changed to 
```DataDict[str(int(RefKey))]```
because RefKey is now being interpreted as a float when it's read from JSON, even though there's no decimal on it anywhere. 

This isn't a difficult change for me to make, but it's a surprise than this applies to all numeric values loaded from JSON.

TheYellowArchitect commented:
`var current_level: int = 12
print(current_level)` would print 12 before this PR, but now prints 12.0
Example before this PR to work:

```gdscript
var current_level: int = 12
	while current_level > target_level:
		level_holder.get_node("Level" + String.num(current_level)).hide()
```
after this PR

```gdscript
var current_level: int = 12
	while current_level > target_level:
		level_holder.get_node("Level" + String.num(current_level).replace(".0","").hide()
```

KoBeWi commented:
> would print 12 before this PR, but now prints 12.0

It doesn't

![image](https://github.com/user-attachments/assets/e3a5987c-330d-47f7-9331-8cac6de68b88)

the number is int.

`String.num()` works with floats. Just use `str()`.

TheYellowArchitect commented:
Thank you for the swift response - was already changing my conversions so I am glad you replied before I finished.
It turns out that just using `str()` instead of `String.num()` doesn't worsen integers :+1:

jerryjrowe commented:
I'm pretty sure this is the right thread---
Integers are getting the extra ".0" on them all over the place, in internal functions, even.

With sign(), specifically, it used to return 1, 0, -1.
Now it returns 1.0, 0, -1.0

And since 1 as an integer isn't the same thing as 1.0, anything that tries to, say, match a value with sign() will not work as intended.

Also, a related weird thing that is definitely happening on Linux...

If you have a variable like `@export var neato_dictionary : Dictionary[String,Variant] {
"thing_one" : 1,
"thing_two" : 1.0,
}`

..if you iterate over the values in the dictionary, it'll recognize the former as an int, and the latter as a float.

But when you use `str(neato_dictionary.values()[0]` (or anything similar to get to that specific value), it'll print "1.0" not "1"

KoBeWi commented:
Note that `sign()` takes and returns Variant, so if you got float, it means your argument is float. Make sure the argument is int or use `signi()` instead.

iegehang commented:
Hello, I needed a feature included only in 4.4 version. So I switched my project. But I'm currently having a problem with this pull feature now. All integer values are returning wrong in WebApi responses. For example; item_id stored as 1 is now 1.0. When I set this responses in a "json" object, all i see is values like 1.0, 2.0 etc. Now I have to convert whole project responses prior to that which is 19k lines of codes. Isn't there an option to change this in a project worldwide?

Calinou commented:
> Edit: I suspected something during serialization; while getting HTTP responses; I am using "JSON.parse_string(body.get_string_from_utf8()" as suggested in official documents. If ".0" appears at the end of integer values just because we are using "get_string_from_utf8" function, this needs to be fixed.

JSON doesn't have integer type, it only has a number type which is parsed as a float in most implementations. Some implementations have a way to force integers to be used when parsing JSON; see https://github.com/godotengine/godot-proposals/issues/8830.

iegehang commented:
> > Edit: I suspected something during serialization; while getting HTTP responses; I am using "JSON.parse_string(body.get_string_from_utf8()" as suggested in official documents. If ".0" appears at the end of integer values just because we are using "get_string_from_utf8" function, this needs to be fixed.
> 
> JSON doesn't have an integer type. Instead, it only has a `number` type which is parsed as a float in most implementations. Some implementations have a way to force integers to be used when parsing JSON; see [godotengine/godot-proposals#8830](https://github.com/godotengine/godot-proposals/issues/8830).

Problem here is you can't deny lots of people using serialization and api results are always the same just because of a .0 extension to an integer value as i never saw before in any other languages. Now all data we have been using requires a deep refactoring. You should have confirmed this pull request after a little thinking about consequences.

Zireael07 commented:
JSON is a standard. It does not have "integers". You seem to have been misusing the standard. This PR's main goal was to make floats vs integers explicit and it surfaced that a lot of folks have been using one when they meant the other.

iegehang commented:
Oh sir, please read what i am telling up there and try to understand my point. JSON has integers or not, that's no matter. One way or another serialization is part of lots of projects. IF a PR effecting a common usage, you have to "think" about the consequences before applying it. This PR is affecting usage of serialization clearly because it seems like anything inside the "get_string_from_utf8" forced out to print ".0". In a very basic response of mine; sql stores the value as 1, WebAPI stores the value as 1. But Godot converts it to 1.0 during serialization. Just why? What would happen if you apply this PR to a language like Javascript. Can you imagine the chaos? Well I took a look at whole issues and repository logs here and I don't understand why Godot developers became so arrogant. Everyone has a meaningless fanaticism over things. Anyway I'm simply saying again; this PR made a method meaningless that has been used frequently.

dalexeev commented:
As said above, the JSON specification does not distinguish between `int` and `float`, there is only `Number`. Regardless of the presence or absence of trailing `.0` in a JSON string obtained using `JSON.stringify()`, after deserializing the string using `JSON.parse_string()` you always get a `float`. This has **not** changed, in 4.3 the same behavior, you can check this with `typeof()` or `var_to_str()`.

```gdscript
@tool
extends EditorScript

func test(value):
    var decoded = JSON.parse_string(JSON.stringify(value))
    prints(value, type_string(typeof(value)),
            "->", decoded, type_string(typeof(decoded)))

func _run():
    test(1)
    test(1.0)
```

4.3:

```
1 int -> 1 float
1 float -> 1 float
```

master:

```
1 int -> 1.0 float
1.0 float -> 1.0 float
```

Although `1 == 1.0` evaluates to `true`, these are values ​​of different types and in many cases they are considered unequal (dictionary keys, `match` patterns, etc.). This change is that other string conversions (`str()`, `print()`, and some others) now make this more obvious. The previous behavior was confusing and misleading (for instance, `print({ 1: "a", 1.0: "b"})` prints `{ 1: "a", 1: "b" }` in 4.3).

I'm sorry to hear that this seems to have been a significant nuisance for you. However, you probably already had hidden bugs due to Godot's non-obvious behavior. It's possible that the bugs didn't show up because of implicit type conversions, but it's too risky and fragile to rely on that. This change just exposed the real problem, the mismatch between your expectations and how it actually works.

Regarding conversions between native and JSON types, we recently added a new API that is intended to be as preserving and unambiguous a conversion as possible. Unfortunately the current implementation has several flaws, so I've opened PR #99765 to fix them.

iegehang commented:
Thank you for your reply sir. Honestly I had upgraded from 4.2 to 4.4. I literally had no idea, 4.3 also had this conversion. Thanks for the #99765

Vetrro commented:
is there a way to toggle the trailing .0 off or an easy way to remove it that I cant think of? in a few areas in my project it doesn't look very nice and I don't know how to fix it with this change. for example in one of my menus there is sliders for volume control and the step size is set to 1 and the text displayed based on my slider values now has a .0 on all of them which I don't like the look of. Is there a way to set sliders to use integers instead of floats for now for min and max value?

KoBeWi commented:
You have to convert the value to int before making string, i.e. `str(int(slider.value))` instead of `str(slider.value)`.

Vetrro commented:
Dang alright was hoping there was a fix all kind of solution gonna have to go through and change all my strings throughout my project, thanks for the reply 😄 just wanted to make sure there wasn't something easier than using str(int(value)) like a tick box option or something

aaronfranke commented:
@Vetrro See https://github.com/godotengine/godot-proposals/issues/11277

Arazalera commented:
How how do I not use this for non integers (1 1.5 2 2.5)?

aaronfranke commented:
@Arazalera See this proposal: https://github.com/godotengine/godot-proposals/issues/11277
